lib/ws-kernel-picker.coffee

```diff
@@ -51,71 +51,85 @@ class WSKernelPicker
             return
 
         @_grammar = grammar
+        @_kernelSpecFilter = kernelSpecFilter
         @_path = atom.workspace.getActiveTextEditor().getPath() + '-' + uuid.v4()
         gatewayListing = new CustomListView('No gateways available', @onGateway.bind this)
         @previouslyFocusedElement = gatewayListing.previouslyFocusedElement
         gatewayListing.setItems gateways
         gatewayListing.setError 'Select a gateway' # TODO(nikita): maybe don't misuse error
 
-    onGateway: (spec) ->
-        console.log('Picked a gateway')
-        sessionListing = new CustomListView('No sessions available', @onSession.bind this)
-        sessionListing.previouslyFocusedElement = @previouslyFocusedElement
-        sessionListing.setLoading 'Loading sessions...'
-
-        services.listRunningSessions(spec.options)
-        .then (sessionModels) ->
-            items = sessionModels.map (model) ->
-                if model.notebook?.path?
-                    name = model.notebook.path
-                else
-                    name = "Session #{model.id}"
-                return {
-                    name: name
-                    model: model
-                    options: spec.options
-                }
-            items.unshift
-                name: '[new session]'
-                model: null
-                options: spec.options
-            sessionListing.setItems(items)
-
-        , (err) =>
-            # Gateways offer the option of never listing sessions, for security
-            # reasons.
-            # Assume this is the case and proceed to creating a new session.
-            @onSession
-                name: '[new session]'
-                model: null
-                options: spec.options
-
-    onSession: (spec) ->
-        unless spec.model?
-            kernelListing = new CustomListView('No kernel specs available', @startSession.bind this)
-            kernelListing.previouslyFocusedElement = @previouslyFocusedElement
-            kernelListing.setLoading 'Loading kernel specs...'
-
-            services.getKernelSpecs(spec.options)
-            .then (specmodels) =>
-                items = _.map specmodels.kernelspecs, (specmodel) =>
-                    options = Object.assign({}, spec.options)
-                    options.kernelName = specmodel.name
-                    options.path = @_path
+    onGateway: (gatewayInfo) ->
+        services.getKernelSpecs(gatewayInfo.options)
+        .then (specmodels) =>
+            kernelSpecs = _.filter specmodels.kernelspecs, (specmodel) =>
```

Comment:
I'm not a native speaker but I think using camel casing for `specModels` and `specModel` would be better. At least it would be consistent with `sessionModels`. What do you think?